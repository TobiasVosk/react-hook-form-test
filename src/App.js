import './App.css';
import { useForm } from 'react-hook-form';
import React from 'react';
import { Button, Card } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css'

function App() {
  const { register, handleSubmit, watch, errors } = useForm();
  const [ submitted, setSubmitted ] = React.useState(false);
  const onSubmit = data => {
      setSubmitted(true)
      console.log(data);
  };

  return (
    <div className="container mt-4">
      <div className="card w-50 mx-auto">
        <div className="card-body">
          <form className="form-group" onSubmit={handleSubmit(onSubmit)}>
            <div>
              <input placeholder="Name" className="form-control mb-3" name="name" defaultValue="test" ref={register} title="name"/>

              <div className="mb-3">
                <input placeholder="Last Name" className={`form-control ${errors.lastName ? "invalid" : ""}`} name="lastName" title="lastName"
                  ref={register({
                    required: {value: true, message: "Last name is required"}, 
                    pattern: {value: /^[a-zA-Z ]+$/i, message: "Name is invalid. Only letters and spaces are accepted"}
                  })} />
                {errors.lastName && <small className="text-danger" title="lastNameValidation">{errors.lastName.message}</small>}
              </div>
              <Button className="btn btn-primary btn-block mb-3" type="submit" title="submitButton">Submit</Button>

              {submitted && <small className="text-success" title="submittedConfirmation">Submitted</small>}
            </div>
          </form>
        </div>
      </div>
    </div>
  );
}

export default App;
