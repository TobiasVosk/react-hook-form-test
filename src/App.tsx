import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css'
import React from 'react';
import {BrowserRouter as Router, Route} from "react-router-dom"
import Form from './form';

function App() {

  return (
    <Router>
      <Route path="" component={Form}></Route>
    </Router>
  );
}

export default App;
