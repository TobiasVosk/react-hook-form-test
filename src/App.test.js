import { act, fireEvent, getByTitle, render, screen } from '@testing-library/react';
import App from './App';
import renderer from 'react-test-renderer';

describe("Form", () => {
  it("Should match the snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe("Name input", () => {
  it("exists in document", () => {
    const { queryByTitle } = render(<App />);
    const nameInput = queryByTitle("name");
    expect(nameInput).toBeInTheDocument();
  })

  it("has the default text", () => {
    const { queryByTitle } = render(<App />);
    const nameInput = queryByTitle("name");
    expect(nameInput.value).toBe("test");
  })

  it("can be written on", () => {
    const { queryByTitle } = render(<App />);
    const nameInput = queryByTitle("name");
    expect(nameInput.value).not.toBe("Changed the value");
    fireEvent.change(nameInput, { target: { value: "Changed the value" } })
    expect(nameInput.value).toBe("Changed the value");
  })
})

describe("Last name input", () => {
  it("exists in document", () => {
    const { queryByTitle } = render(<App />);
    const lastNameInput = queryByTitle("lastName");
    expect(lastNameInput).toBeInTheDocument();
  })

  it("can be written on", () => {
    const { queryByTitle } = render(<App />);
    const lastNameInput = queryByTitle("lastName");
    expect(lastNameInput.value).not.toBe("Changed the value");
    fireEvent.change(lastNameInput, { target: { value: "Changed the value" } })
    expect(lastNameInput.value).toBe("Changed the value");
  })
})

describe("Submit button", () => {
  it("exists in document", () => {
    const { queryByTitle } = render(<App />);
    const submit = queryByTitle("submitButton");
    expect(submit).toBeInTheDocument();
  })

  it("Is not submitted if lastName is empty", async () => {
    const { queryByTitle } = render(<App />);
    const submit = screen.getByText('Submit');

    await act(async () => {
      fireEvent.click(submit);
    });
    const lastNameValidation = queryByTitle("lastNameValidation");
    expect(screen.getByText('Last name is required')).toBeInTheDocument();
    expect(lastNameValidation).toBeInTheDocument();
  })

  it("Is not submitted if lastName has symbols", async () => {
    const { queryByTitle } = render(<App />);
    const submit = screen.getByText('Submit');

    const lastNameInput = queryByTitle("lastName");
    fireEvent.change(lastNameInput, { target: { value: "@" } })

    await act(async () => {
      fireEvent.click(submit);
    });
    expect(screen.getByText('Name is invalid. Only letters and spaces are accepted')).toBeInTheDocument();
    expect(queryByTitle('submittedConfirmation')).not.toBeInTheDocument();
  })

  it("Is submitted if lastName has letters and spaces and name is changed", async () => {
    const { queryByTitle } = render(<App />);
    const submit = screen.getByText('Submit');

    const nameInput = queryByTitle("name");
    fireEvent.change(nameInput, { target: { value: "This is a wonderful name" } })

    const lastNameInput = queryByTitle("lastName");
    fireEvent.change(lastNameInput, { target: { value: "This is a wonderful last name" } })

    await act(async () => {
      fireEvent.click(submit);
    });

    expect(queryByTitle('lastNameValidation')).not.toBeInTheDocument();
    expect(queryByTitle('submittedConfirmation')).toBeInTheDocument();
  })
})