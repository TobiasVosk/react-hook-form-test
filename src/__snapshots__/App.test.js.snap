// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Should match the snapshot 1`] = `
<div
  className="container mt-4"
>
  <div
    className="card w-50 mx-auto"
  >
    <div
      className="card-body"
    >
      <form
        className="form-group"
        onSubmit={[Function]}
      >
        <div>
          <input
            className="form-control mb-3"
            defaultValue="test"
            name="name"
            placeholder="Name"
            title="name"
          />
          <div
            className="mb-3"
          >
            <input
              aria-invalid="false"
              className="form-control "
              name="lastName"
              placeholder="Last Name"
              title="lastName"
            />
          </div>
          <button
            className="btn btn-primary btn-block mb-3 btn btn-primary"
            disabled={false}
            title="submitButton"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
